Ladybird/MachPortServer.cpp

```diff
@@ -68,23 +68,38 @@ void MachPortServer::thread_loop()
             break;
         }
 
-        if (message.header.msgh_id != Core::Platform::SELF_TASK_PORT_MESSAGE_ID) {
-            dbgln("Received message with id {}, ignoring", message.header.msgh_id);
+        if (message.header.msgh_id == Core::Platform::BACKING_STORE_IOSURFACES_MESSAGE_ID) {
+            auto pid = static_cast<pid_t>(message.body.parent_iosurface.trailer.msgh_audit.val[5]);
+            auto const& backing_stores_message = message.body.parent_iosurface;
+            auto front_child_port = Core::MachPort::adopt_right(backing_stores_message.front_descriptor.name, Core::MachPort::PortRight::Send);
+            auto back_child_port = Core::MachPort::adopt_right(backing_stores_message.back_descriptor.name, Core::MachPort::PortRight::Send);
+            auto* metadata = reinterpret_cast<Core::Platform::BackingStoreMetadata*>(backing_stores_message.metadata.address);
+            if (on_receive_backing_stores)
+                on_receive_backing_stores({ .pid = pid,
+                    .page_id = metadata->page_id,
+                    .front_backing_store_id = metadata->front_backing_store_id,
+                    .back_backing_store_id = metadata->back_backing_store_id,
+                    .front_backing_store_port = move(front_child_port),
+                    .back_backing_store_port = move(back_child_port) });
             continue;
         }
 
-        if (MACH_MSGH_BITS_LOCAL(message.header.msgh_bits) != MACH_MSG_TYPE_MOVE_SEND) {
-            dbgln("Received message with invalid local port rights {}, ignoring", MACH_MSGH_BITS_LOCAL(message.header.msgh_bits));
+        if (message.header.msgh_id == Core::Platform::SELF_TASK_PORT_MESSAGE_ID) {
+            if (MACH_MSGH_BITS_LOCAL(message.header.msgh_bits) != MACH_MSG_TYPE_MOVE_SEND) {
+                dbgln("Received message with invalid local port rights {}, ignoring", MACH_MSGH_BITS_LOCAL(message.header.msgh_bits));
+                continue;
+            }
+
+            auto const& task_port_message = message.body.parent;
+            auto pid = static_cast<pid_t>(task_port_message.trailer.msgh_audit.val[5]);
+            auto child_port = Core::MachPort::adopt_right(task_port_message.port_descriptor.name, Core::MachPort::PortRight::Send);
```

Comment:
I forget my previous implementation, but what happens if the right that's in the port descriptor doesn't have the right you think it does? Do we assert? or otherwise complain?